app/models/ems_refresh.rb

```diff
@@ -140,31 +146,61 @@ def self.get_ar_objects(target, single_id = nil)
     end
   end
 
-  def self.queue_merge(targets, ems, sync = false)
-    sync ? queue_merge_sync(targets, ems) : queue_merge_async(targets, ems)
-  end
+  def self.queue_merge(targets, ems, create_task = false)
+    task = create_refresh_task(ems, targets) if create_task
 
-  def self.queue_merge_async(targets, ems)
-    # Items will be naturally serialized since there is a dedicated worker.
-    MiqQueue.put_or_update(
+    queue_options = {
       :queue_name  => MiqEmsRefreshWorker.queue_name_for_ems(ems),
       :class_name  => name,
       :method_name => 'refresh',
       :role        => "ems_inventory",
-      :zone        => ems.my_zone
-    ) do |msg, item|
+      :zone        => ems.my_zone,
+    }
+
+    # If this is the only refresh then we will use the task we just created,
+    # if we merge with another queue item then we will return its task_id
+    task_id = nil
+
+    # Items will be naturally serialized since there is a dedicated worker.
+    MiqQueue.put_or_update(queue_options) do |msg, item|
       targets = msg.nil? ? targets : (msg.args[0] | targets)
+      task_id = msg && msg.task_id ? msg.task_id.to_i : task.try(:id)
+
       item.merge(
-        :args        => [targets],
-        :msg_timeout => queue_timeout,
-        :task_id     => nil)
+        :args         => [targets],
+        :task_id      => task_id,
+        :msg_timeout  => queue_timeout,
+        :miq_callback => {
+          :class_name  => task.class.name,
+          :method_name => :queue_callback,
+          :instance_id => task_id,
+          :args        => ['Finished']
+        }
+      )
     end
+
+    # If we merged with an existing queue item we don't need a new
+    # task, just use the original one
+    task.delete if task && task_id != task.id
```

Comment:
Oh, it's here...wonder if there's a nicer way to do this so as not to create then destroy it.